warcio/limitreader.py

```diff
@@ -67,3 +74,144 @@ def wrap_stream(stream, content_length):
 
         return stream
 
+
+# ============================================================================
+class DigestVerifyingReader(object):
+    """
+    A reader which verifies the digest of the wrapped reader
+    """
+
+    def __init__(self, stream, length, check_digests=False, record_type=None,
+                 payload_digest=None, block_digest=None, segment_number=None):
+
+        self.stream = stream
+        self.limit = length
+        if check_digests:
+            self.exception = ArchiveLoadFailed
+        else:
+            self.exception = None
+
+        if record_type == 'revisit':
+            block_digest = None  # XXX my bug, or is example.warc wrong?
+            payload_digest = None  # no payload, so can't check it
+        if segment_number is not None:  #pragma: no cover
+            payload_digest = None
+
+        self.payload_digest = payload_digest
+        self.block_digest = block_digest
+
+        self.payload_digester = None
+        self.payload_digester_obj = None
+        self.block_digester = None
+
+        if block_digest:
+            try:
+                algo, _ = _parse_digest(block_digest)
+                self.block_digester = Digester(algo)
+            except ValueError:
+                self.problem('unknown hash algorithm name in block digest')
+                self.block_digester = None
+        if payload_digest:
+            # if these are going to raise, have them do it here
+            try:
+                algo, _ = _parse_digest(self.payload_digest)
+                self.payload_digester_obj = Digester(algo)
+            except ValueError:
+                self.problem('unknown hash algorithm name in payload digest')
+
+    def begin_payload(self):
+        self.payload_digester = self.payload_digester_obj
+        if self.limit == 0:
+            # payload is of length 0
+            if not _compare_digest_rfc_3548(self.payload_digester, self.payload_digest):
+                self.problem('payload digest failed: {}'.format(self.payload_digest))
+                self.payload_digester = None  # prevent double-fire
+
+    def _update(self, buff):
```

Comment:
RE my why does `DigestVerifyingReader` not inherit from `LimitReader`
```suggestion
    def _update(self, buff):
        buff = super(DigestVerifyingReader, self)._update(buff)
        # digest checks
```

Reply:
I need to know "limit" for the subsequent "if self.limit == 0" check, and LimitReader doesn't make it possible to read the limit. So I either have to duplicate 2 lines of code (my current choice) or futz with LimitReader. I don't know a clean way to alter LimitReader.

Reply:
limit is a instance property of limit reader.

https://github.com/webrecorder/warcio/blob/c64c4394805e13256695f51af072c95389397ee9/warcio/limitreader.py#L9

If you were to inherit from it you would be able to use it here :)

Reply:
Uh, OK, I thought that broke encapsulation, but I'm not a Python expert.